### ISSUE TITLE
Vulnerable regexp in rule 932140

### ISSUE THREAD

**Issue by s0md3v** [SpiderLabs/owasp-modsecurity-crs#1354]
The vulnerable regular expression is located in
`/crs/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf`
on line 404. [[Link](https://github.com/SpiderLabs/owasp-modsecurity-crs/blob/d0440071931e0d56256348334987bd38403d33c0/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf#L404)]

The vulnerability is caused by nested repetition operators and can be exploited with the following string

```
for/d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d/ %x in(x) dx
```

**Comment by theMiddleBlue** [comment]
Unlike what you reported on #1359 I'm not able to reproduce the issue. Any help on defining a ReDoS HTTP request that matches 932140?

**Comment by spartantri** [comment]
Hi @s0md3v ,there is a limit for following those bad regex matches in modsecurity, have you tested this in modsecurity? what is your current modsecurity setup and what is the extent of the impact you noticed?

[PCRE match limit](https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#SecPcreMatchLimit)
[PCRE match limit recursion](https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#SecPcreMatchLimitRecursion)

**Comment by s0md3v** [comment]
Hi @spartantri ,

I don't have modsecurity setup on my machine and I don't really need one because the vulnerability exists in the regular expression.

I simply extracted all the regular expressions from the configuration files, ran a quick grep to find the find the potentially vulnerable ones, tried to exploit them one by one by using [regex101.com](http://regex101.com/) and then verified the impact using `re.search(pattern, exploit_string)` in Python.
I apologize for not using `match` instead of `search` which is the reason 4 out 5 vulnerabilities were not reproducible.

Also, I wasn't aware about the limits but I have a few questions about this implementation
1. The timeout seems to be 1.5 seconds but what happens if the matching times out? Is the request blocked?
2. Shouldn't we fix the root cause instead of relying on such limits? PS: I am not a regex wizard but I believe the recursion limit has nothing to with backtracking which is the real culprit.

I believe the regular expressions should be fixed but it's up to your team to make the final decision.

**Comment by spartantri** [comment]
Unless the configuration is made purposely changed to extremely high limits this is not a high DoS vulnerability as the title suggest as the limits are like adding an {#} to the regex in place of a recursive match.

When limits are reached, depending on your particular setup and the rule logic the transaction most probably will be blocked. I agree there are some regex that need some love to make them more efficient and this fixes should definitively be implemented to prevent a possible vulnerability when an administrator sets crazy values on their setup.

Maybe we can add this to CONTRIBUTING as guidelines to keep this in mind and prevent this from happening in the future, and also tune and benchmark the CRS for performance.

**Comment by s0md3v** [comment]
Thanks for making this clear but what's the default behavior?
Does mod_security flags the request as malicious when the timeout value is reached?

**Comment by dune73** [comment]
This issue is being referenced as [CVE-2019-11388](https://nvd.nist.gov/vuln/detail/CVE-2019-11388) by NIST.

I confirm that this issue is not present with ModSecurity.

I have tried to reproduce the behaviour with the following call:

```
curl -v "http://localhost" -d "x=for/d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d/ %x in(x) dx"
```

against the following setups:
* Vanilla ModSec 2.9.3 on Apache
* ModSec 2.9.3 on Apache with PCRE Match Limits of 50M
* ModSec 3.0.3 on Nginx 1.3.12

None of them shows the behaviour claimed in the issue.

@s0md3v: Please explain.

**Comment by dune73** [comment]
This issue is being referenced as [CVE-2019-11388](https://nvd.nist.gov/vuln/detail/CVE-2019-11388) by NIST.

**Comment by s0md3v** [comment]
I did not craft exploits for the whole strings because they could have been abused. I wrote exploits for the vulnerable sub-patterns.

**Comment by spartantri** [comment]
@dune73  Some engines (like the one of regex101 which is different than modsec) get catastrophic backtracking errors with
```for/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d/d %variable in(set) do command```
But wasn't able to reproduce or had delays in modsec v2 with Apache

**Comment by s0md3v** [comment]
I tested it with Python.

**Comment by fgsch** [comment]
From pcre:

```
  re> "for(?:\/[dflr].*)* %+[^ ]+ in\(.*\)\s?do"
data> for/d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d//d/ %x in(x) dx
Error -8 (match limit exceeded)
```

**Comment by dune73** [comment]
Thanks for your feedback @s0md3v. I see the problem in the regex and I can confirm what @fgsch got with `pcre2test`. Yet I am not sure if you can really exploit this via CRS/ModSec. After all CRS rule 932140 applies two transformations to the payload and I do not see any misbehaviour of the webserver with your subpattern.

Also: Where would the subpattern and the pattern lead to different behaviours of the ModSec engine? I do not understand this completely yet.

**Comment by fgsch** [comment]
We should correct the regexp regardless since the parameter can only appear once.
I've commented on the PR, and hopefully after it's updated we can merge it if there are no objections.

**Comment by fgsch** [comment]
Addressed in #1472
